sdk/servicebus/Azure.Messaging.ServiceBus/api/Azure.Messaging.ServiceBus.netstandard2.0.cs

```diff
@@ -94,6 +95,8 @@ public partial class ServiceBusClient : System.IAsyncDisposable
         public virtual System.Threading.Tasks.Task<Azure.Messaging.ServiceBus.ServiceBusSessionReceiver> CreateSessionReceiverAsync(string queueName, Azure.Messaging.ServiceBus.ServiceBusSessionReceiverOptions options = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
         public virtual System.Threading.Tasks.Task<Azure.Messaging.ServiceBus.ServiceBusSessionReceiver> CreateSessionReceiverAsync(string topicName, string subscriptionName, Azure.Messaging.ServiceBus.ServiceBusSessionReceiverOptions options = null, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
         public virtual System.Threading.Tasks.ValueTask DisposeAsync() { throw null; }
+        public System.Collections.ObjectModel.ReadOnlyCollection<Azure.Messaging.ServiceBus.Plugins.ServiceBusPlugin> GetPlugins() { throw null; }
```

Comment:
Do we support changing clients during client lifetime? Any reason not to have them on options?

Reply:
This feature seems very similar to HTTP pipeline policies and the departure from the pattern is unfortunate. It also opens the door for thread safety issues and customer confusion around when are clients applied.

Reply:
Agreed.   I was a fan of Josh's proposed `PluginRegistry` type idea (from the linked issue) that had all the sugar for registering and served as a spot to attach extension methods which could then be passed as part of the options and would be frozen once the client was created.

Reply:
The concern with using PluginOptions was the amount of boiler plate code:https://github.com/Azure/azure-sdk-for-net/issues/11986#issuecomment-645666379

Regarding thread safety, the plugins are copied on creation of the sender/receiver.

Reply:
To quote the comment:

**Poor**

```c#
var pluginOptions = new ServiceBusPluginOptions();
pluginOptions.RegisterPlugin(new TestPlugin());
var clientOptions = new ServiceBusClientOptions
{
    PluginOptions = pluginOptions
};
var client = new ServiceBusClient(connStr, clientOptions);
```

**Optimal**

```c#
var client = new ServiceBusClient(connStr, clientOptions);
client.RegisterPlugin(new TestPlugin());

// alternatively, extension methods provided by the plugin author 
// to reduce implementation details and breaking change releases

client.UsePluginX(/*optional plugin configuration object*/);
```

What about the following:

***
var clientOptions = new ServiceBusClientOptions();
clientOptions .RegisterPlugin(new TestPlugin());
var client = new ServiceBusClient(connStr, clientOptions);
```

Reply:
Yes, but if I call an extension method on the client, the client instance is passed in.   That extension method may perform work on another thread and then mutate the client state.   Since we can't offer guarantees about how the plugins may choose to register themselves, we could potentially have a hidden race condition if this isn't synchronized.

Reply:
@SeanFeldman thoughts on @pakrym suggestion?

Reply:
The proposed option is better than the original API.

One concern I have with the approach is that it will require constructing and using `ServiceBusClientOptions`. Customers that do not wish to change the client defaults will be forced to add it to be able to use plugins. `ServiceBusClientOptions` will be used only when the transport cannot be AMQP, there's a proxy that needs to be configured or the default retry policy is not suitable. I.e. the `ServiceBusClient` is always there and the extension methods targeting those would be showing up in the intellisense right away. With the newly proposed suggestion, intellisense will not help, hindering discoverability. Which is one of the Azure SDK tenants, is it not?

So it's discoverability/usability vs compliance with the pattern.

Reply:
Agreed that it is slightly less discoverable, but I see the smaller surface area on the client as a benefit as well as the consistency with other libraries. It is also nice to have the client be immutable.

Reply:
@SeanFeldman: I think that we're in agreement that it would be a friendlier pattern on the client itself.   The underlying problem that we ran into is that it creates some race conditions if you can register plugins while there are active Senders/Receivers.   

Getting around this is easy enough, if you clone the set and pass that frozen copy when you create a Sender/Receiver.   However, that makes it potentially difficult to understand which plugins are available to which Sender/Receiver.  It also creates the unintuitive pattern that if I create a new Sender/Receiver using the client, my new instance may have a different set of plugins than the Sender/Receiver that I had previously created.

Ensuring the set of plugins is immutable when the client is constructed seemed the lesser of evils when considered against the above or a pattern like "call BeginPluginRegistration when you start and EndPluginRegistration when you finish", which would seem to open a bunch of other invariants like "you can't register plugins for a client after a Sender/Receiver is created by it for the first time"  or "you can't create a Sender/Receiver once BeginPluginRegistration has been called until the end call is observed".    I think that explosion of client states makes it far more difficult to reason about the client.

It's possible that we're overlooking some better ideas for a pattern there.   Any thoughts that you have for a design consideration would be welcome.

Reply:
> I think that we're in agreement that it would be a friendlier pattern on the client itself.

That's good to hear.

> Any thoughts that you have for a design consideration would be welcome.

There are certain assumptions that can and probably should be made. For example, the case for "you can't register plugins for a client after a Sender/Receiver is created by it for the first time" is a no brainer. End-users will always register their plugins that would be applicable for all senders/receivers and do it early on. That is why I was wondering if it should be on the `ServiceBusClient` surface rather than individual senders and receivers. I understand that an options object is designed with the flexibility in mind to allow each sender/receiver it's own nuances, but with ASB plugins end-users mostly wanted it on the "global pipeline".

We can discuss this in details on the call.

Reply:
>I think that we're in agreement that it would be a friendlier pattern on the client itself.

I wouldn't necessarily say that more discoverable always means more user-friendly, particularly when we are talking about a feature that is not likely to be needed for the majority of use cases.